pkg/pushsync/pushsync.go

```diff
@@ -77,59 +78,153 @@ func (ps *PushSync) Close() error {
 	return nil
 }
 
-// handler handles chunk delivery from other node and inserts it to localstore.
-// it also sends this chunk to the closest peer if one exists.
+// handler handles chunk delivery from other node and forwards to its destination node.
+// If the current node is the destination, it stores in the local store and sends a receipt.
 func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
-	_, r := protobuf.NewWriterAndReader(stream)
+	w, r := protobuf.NewWriterAndReader(stream)
 	defer stream.Close()
 
-	var ch pb.Delivery
-
-	if err := r.ReadMsg(&ch); err != nil {
-		if err == io.EOF {
-			return nil
-		}
-		return err
+	// Get the delivery
+	ch, err := ps.getChunkDelivery(r)
+	if err != nil {
+		return fmt.Errorf("chunk delivery: %w", err)
 	}
 
-	// create chunk and store it in the local store
+	// create chunk
 	addr := swarm.NewAddress(ch.Address)
 	chunk := swarm.NewChunk(addr, ch.Data)
-	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
-	if err != nil {
-		return err
-	}
 
-	// push this to the closest node too
-	peer, err := ps.peerSuggester.ClosestPeer(addr)
+	// Select the closest peer to forward the chunk
+	peer, err := ps.peerSuggester.ClosestPeer(chunk.Address())
 	if err != nil {
+		// If i am the closest peer then store the chunk and send receipt
 		if errors.Is(err, topology.ErrWantSelf) {
-			// i'm the closest - nothing to do
+
+			// Store the chunk in the local store
+			_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+			if err != nil {
+				return fmt.Errorf("chunk store: %w", err)
+			}
+			ps.metrics.TotalChunksStoredInDB.Inc()
+
+			// Send a receipt immediately once the storage of the chunk is successfull
+			err = ps.sendReceipt(w, chunk.Address())
```

Comment:
you can probably `return ps.sendReceipt(...)` and remove the error check and nil return

Reply:
Same as above

Reply:
This is where i meant to say the above comment.
What if we get error sending to the stream or encoding in the protobuf?

Reply:
misunderstood the comment... got it done..